Add new docstrings for most functions and methods in vd.py

DataBranner opened the pull request.

brannerchinese committed:
Add docstrings for main classes and some methods in addons/

brannerchinese committed:
Give docstrings more uniform punctuation and fix one other typo

brannerchinese committed:
Revise addon docstrings as per @saulpw's comments

brannerchinese committed:
Add or revise docstrings for the majority of functions and methods.

brannerchinese committed:
Add .gitinore and add one docstring to bin/vd

brannerchinese committed:
Revise a few of the new docstrings

saulpw committed:
Merge branch 'develop' into dpb-docstrings-vd.py

saulpw commented on visidata/addons/gsheets.py:
```diff
@@ -49,4 +55,4 @@ def open_gsheet(p):
 #    vs.command(Key.ENTER, 'cursorRow')
 #    return vs
-
+h_pyobj
```
jetsam?

DataBranner replied:
Deleted.

DataBranner commented:
@saulpw: The current pull request offers docstrings for the majority of functions and methods in vd.py and the addons.

saulpw commented on visidata/vd.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 
-'VisiData core functionality'
+"""Provide VisiData core functionality"""
```
I'd prefer without "Provide", here and below.  I understand the rationale for a common phrase structure within git commits and issue titles, but I don't think that applies here.

DataBranner replied:
Done in a few cases, producing nominal docstrings.

saulpw commented on visidata/vd.py:
```diff
@@ -299,23 +302,27 @@ def tryType(T, v):
 windowHeight = 0
 
 def joinSheetnames(*sheetnames):
+    """Concatenate sheet names using separator ("joiner") from options."""
```
Use concrete name (`sheetname_joiner`), especially if within quotes.  `options.sheetname_joiner` seems even clearer to me, since that's how it's used in the code.

DataBranner replied:
Done, within 79 characters.

saulpw commented on visidata/vd.py:
```diff
@@ -299,23 +302,27 @@ def tryType(T, v):
 
 def enumPivot(L, pivotIdx):
-    'like enumerate() but starts after pivotIdx and wraps around to end at pivotIdx'
+    """Model Python `enumerate()` but starting midway through sequence `L`.
```
This one is a good upgrade overall, but loses content.  Notably the enumeration starts *after* pivotIdx, and finally ends on pivotIdx.

DataBranner replied:
Done.

saulpw commented on visidata/vd.py:
```diff
@@ -325,20 +332,30 @@ def enumPivot(L, pivotIdx):
-# A .. Z AA AB ...
+# A .. Z AA AB .. ZZ
 defaultColNames = list(itertools.chain(string.ascii_uppercase, [''.join(i) for i in itertools.product(string.ascii_uppercase, repeat=2)]))
+# TODO: How about this, more concise and more readable:
```
Yes, please!  Great improvement.

saulpw commented on visidata/vd.py:
```diff
@@ -400,14 +422,17 @@ def getkeystroke(self):
                     return True
             return False
+        # TODO: I believe we can more concisely replace function body with:
+        #   return any([c.getDisplayValue(row) for c in columns])
```
`any([func(c.getDisplayValue`, but yes, I agree. Good suggestion.

DataBranner replied:
My TODO note corrected; I will deal with all TODOs separately after docstrings have been merged.

saulpw commented on visidata/vd.py:
```diff
@@ -449,7 +474,7 @@ def columnsMatch(sheet, row, columns, func):
                 if moveCursor:
                     sheet.cursorRowIndex = r
-                    status('search wrapped')   # the only reason for the duplicate code block
+                    status('search wrapped')   # the only reason for the duplicate code block TODO: this can be condensed.
```
Don't add TODOs in the source code; prefer that we create issues tagged 'code cleanup'.

DataBranner replied:
Agreed, as per our discussion earlier today. Will go through code again and deal with TODOs after docstrings are merged.

saulpw commented on visidata/vd.py:
```diff
@@ -634,24 +673,31 @@ def thread_profileCode(task, func, *args, **kwargs):
 def ctype_async_raise(thread_obj, exception):
+    """Raise exception for threads running asynchronously.
+
+    Ctypes call follows https://gist.github.com/liuw/2407154."""
```
This is best as a comment immediately before the ctypes call itself.

DataBranner replied:
Done.

saulpw commented on visidata/vd.py:
```diff
@@ -896,9 +977,11 @@ def toggle(self, rows):
         self._selectedRows[id(row)] = row
 
     def unselectRow(self, row):
+        """Unselect given row, if selected; return boolean."""
```
"return True if previously selected"

DataBranner replied:
Done.

saulpw commented on visidata/vd.py:
```diff
@@ -1117,9 +1224,11 @@ def drawColHeader(self, vcolidx):
         return self.visibleCols[vcolidx] in self.keyCols
 
     def draw(self, scr):
+        """Draw given screen object."""
```
This draws the entire screen onto the `scr` curses object.

DataBranner replied:
Done.

saulpw commented on visidata/vd.py:
```diff
@@ -1201,34 +1313,40 @@ def editCell(self, vcolidx=None, rowidx=None):
 
 _sum = sum
+
 def sum(values):
```
This is because builtins can't have .type assigned to them.  This whole aggregator setup needs to be fleshed out.  (I know it's a bit of a mess right now).

DataBranner replied:
No change.

saulpw commented on visidata/vd.py:
```diff
@@ -1910,16 +2110,19 @@ def run(sheetlist=[]):
     return vd().run(_scr)
 
-g_globals = None
+g_globals = None # TODO: If we expect a dict, why not `dict()` here?
```
Workaround for Python module scoping semantics.  There may be a better way, but this is how I was able to access all of the globals from all of the files imported in __init__ (see bottom of that file).

DataBranner replied:
No change. I just mean that assigning `g_globals = dict()` will still be `False` (the same as `g_globals = None`), but the reader will know that a dict is intended.

DataBranner reviewed (commented):
Changes have been made to the docstrings wherever suggested. TODOs have been left in place for separate handling later as discrete pull requests.

brannerchinese committed:
Make revisions to docstrings, following @saulpw's comments.

TODOs have been left in place, and will later be considered for
implementation as individual pull requests, as per a face-to-face
discussion today.

brannerchinese committed:
Remove all TODO items.

saulpw committed:
Merge branch 'develop' into dpb-docstrings-vd.py
